### PR TITLE
Enhanced json normalize

### LINKED ISSUE #23843: Configurable json_normalize with respect to number of levels and Keys to be flattened
#### Code Sample, a copy-pastable example if possible

```python
data = [{"CreatedBy":{"Name": "User001"}, 
        "Lookup":{"TextField":"Some text", "UserField":{"Id":"ID001", "Name": "Name001"}},
        "Image":{"a":"b"}}]
json_normalize(data).to_json(orient="records")
```
#### Current Output
```
[{"CreatedBy.Name":"User001","Image.a":"b",
"Lookup.TextField":"Some text",
"Lookup.UserField.Id":"ID001","Lookup.UserField.Name":"Name001"}]
```
#### Problem description

I want to flatten only specific keys and up to a specific a specific level. 
For eg., I want to flatten until level 1 and skip key image.

#### Expected Output

```
[{"CreatedBy.Name":"User001",
"Image:{a":"b}",
"Lookup.TextField":"Some text",
"Lookup.UserField":{Id":"ID001",
"Name":"Name001"}]
```

#### Output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
pandas: 0.23.4
pytest: None
pip: 18.1
setuptools: 39.1.0
Cython: None
numpy: 1.15.1
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by bhavaniravi**
closes #23843
closes #23861
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/json/normalize.py
Comment (gfyoung) [inline pandas/io/json/normalize.py]: * Use parentheses to avoid using the backslash.
* I think the comment can be more expressive than single-lined bullets.  Can you write something a little more explaining the behavior here?

File: pandas/tests/io/json/test_normalize.py
Comment (gfyoung) [inline pandas/tests/io/json/test_normalize.py]: Can we test multiple `max_level` inputs here.

**Comment by gfyoung** [comment]
Good start!  Can you also add a mini-section to the `whatsnew` explaining what these parameters do and how you use them?

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Did you by chance already look at the `_recursive_extract` function in `json_normalize`? It has a `level` parameter already.

Haven't stepped through the code in detail to see but am curious if there's not a better way of going about this then adding a `max_level` parameter here; it's not very intuitive what the difference between `level` and `max_level` would be with these functions

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: Will it be fine if I change it to `start_level` and `end_level`?

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: To avoid specific keys from getting normalized.

Comment (WillAyd) [reply pandas/io/json/normalize.py]: What is the difference between this and `record_path` in `json_normalize` then?

File: pandas/io/json/normalize.py
Comment (WillAyd) [reply pandas/io/json/normalize.py]: Not sure that's the right solution; my general concern with this PR is that it doesn't make the API that much cleaner for these funcs.

Can you add some tests that interact with the already existing keywords? Maybe I'm just not thinking about it correctly. That would help from an understanding perspective as well as ensuring better regression testing

**Comment by jreback** [comment]
closing. if you want to continue, pls ping. needs to merge master and update to comments.

**Comment by bhavaniravi** [comment]
Yes I want to continue this to merging. I was waiting for #22706 to get merged.

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: `record_path` defines the "path to the data to be normalized". Where as with `max_level` it assumes `record_path` level as 0 and normalizes it until max_level. The key path in `ignore_keys` will be left out

File: pandas/tests/io/json/test_normalize.py
Comment (bhavaniravi) [inline pandas/tests/io/json/test_normalize.py]: @WillAyd Added a test case with existing `record_path` and `meta` keys with newly added `max_level` and `ignore_keys` param

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Can you run `scripts/validate_docstring.py` against this? Should give you some failures to correct

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Unless there's a reason for this to be `None` probably simplifies code if this is 0, no?

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Similar comment here maybe just an empty list yields cleaner code

File: pandas/tests/io/json/test_normalize.py
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: Can you simplify this test via parametrization? Probably OK to parametrize arguments of `max_level` and `expected`

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: `None` for the case where you don't set a `max_level` and the nesting happens at all levels.

**Comment by bhavaniravi** [comment]
@WillAyd. The test passes in local and fails in Python 3.5 of the build.

image

If the change the order of columns it fails in local.  Can you help me fix this?

**Comment by gfyoung** [comment]
This could be indicative of the (somewhat) unpredictable nature of column ordering when passing in a `dict` mapping of values into `DataFrame`.  I bet if you look at the data, you'll see that the column-data mappings are correct, just that ordering is not consistent.

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Change these to 0.25

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Minor capitalization issue here

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: After giving this some more thought...can we split this PR up? I think this one should just focus on `max_level` and a subsequent PR can do `ignore_keys`. Will make review and communication out easier

File: pandas/tests/io/json/test_normalize.py
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: If you only plan on using these in once function then just define there; no need for fixture

File: pandas/tests/io/json/test_normalize.py
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: Does this test focus on `record_path` or `max_level`? Assume the latter from change but unclear from naming

File: pandas/tests/io/json/test_normalize.py
Comment (WillAyd) [inline pandas/tests/io/json/test_normalize.py]: If there is a way to parametrize this with two tests above would be preferable

**Review by WillAyd — CHANGES_REQUESTED** [review]
As mentioned in one comment I think it will be easier if we just focus on one feature at a time here, so if you could orient PR to just do `max_level` here would be easiest for review going forward

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: ``ignore_keys=None`` is the idiom. you can assign an empty list if you want *inside* the function

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: max_level: int, optional
    comment goes here

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: same format as below

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: same

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: no -> number

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: 0.25.0

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
Can you merge master and address comments?

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: you have an extra file above, doc/test.parquet pls remove

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by bhavaniravi** [comment]
Sure will do the changes and give a PR.

**Comment by bpben** [comment]
I wanted to add on to this from some discussion over in #21537 that this fix doesn't handle the case of nested records.

For example:

data = {'a': 1, 'b': 2, 'c': {'d': [{'a': 1}, {'b': 2}], 'e': [{'a': 1}, {'b': 2}]}}
pd.io.json.json_normalize(data, record_path=['c'])

Produces:

0 | 0
-- | --
0 | d
1 | e

Where (ideally) it should produce:

0 | d.a | d.b | e.a | e.b
-- | -- | -- | -- | --
0 | 1 | 2 | 1 | 2

This is mainly because of the following bit:

https://github.com/pandas-dev/pandas/blob/fca2a2752dac842eb11e08aa7a35c42b4eefbc84/pandas/io/json/normalize.py#L287-L290

Which only extracts keys from the records, rather than iterating over the items.  Does this seem within scope for this PR? I can try and add-on here, if that's helpful.

**Comment by jreback** [comment]
if you can merge master and update to comments

**Comment by bpben** [comment]
Happy to help make the updates if @bhavaniravi doesn't mind.  But I'd also be interested in adding a piece to account for the nested record structure.  Any suggestions there? I'm thinking adding a check along the lines of:

```
if isinstance(recs, dict):
     recs = nested_to_record(recs)
```

I have to check if this works, but wanted to check if this should be part of a separate PR or if it's in scope here.

**Comment by WillAyd** [comment]
Closing as stale but ping if you'd like to continue. Otherwise @bpben if you'd like to pick this up in another PR then by all means!

**Comment by bhavaniravi** [comment]
Sorry guys, A bit occupied. Can close it off by End of this week, If that
works.

regards
Bhavani Ravi :)
bhavaniravi.com <https://medium.com/@bhavaniravi>

**Comment by WillAyd** [comment]
No worries. Typically there’s no rush with these just wanted to make sure it wasn’t abandoned. I’ve reopened - thanks for the help!

**Comment by bpben** [comment]
@WillAyd any thoughts on whether my proposed edit might be in scope here or should I open another PR?

**Comment by WillAyd** [comment]
@bpben another PR - I already think the scope of this one might be too large to begin with so breaking up makes the most sense

**Comment by jreback** [comment]
pls merge master

**Comment by WillAyd** [comment]
Can you check test failures? Looks like your last commits also picked up unwanted files

**Comment by bhavaniravi** [comment]
I am trying to fix the test failures. `test_missing_meta` failing. I printed the outputs and found `nan` and `NaN` are not equal. Not sure if `assert_frame_equal` will take care of this.

image

**Comment by WillAyd** [comment]
It looks like one of your values is the literal string nan whereas the other is the actual sentinel. Hope that helps will need to be fixed.

**Comment by bhavaniravi** [comment]
Line number 328 is where I am having the issue 
```
meta_values_ = np.array(v).repeat(lengths)
```
I'm printing v, and the result from above line
```
meta_values ===             ['Alice' 'nan']
type of real value v ===        <class 'float'>
np.array(v).repeat([lengths]) ===  <class 'numpy.str_'>
```

**Comment by jreback** [comment]
you have included get-pip somehow

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: can you make this a proper doc-string (and edit the others), e.g. for level
```
level: int, optional, default 0
     the number of levels......
```

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: this is not lined up

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: is this getting mutated?

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: see my comments above

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: this won't pass doc-strings, I guess its not being checked? also you have mixed quoting, is this an actual result?

@WillAyd

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: why is there a period here?

File: pandas/tests/io/json/test_normalize.py
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: this should be a fixture as well

File: pandas/tests/io/json/test_normalize.py
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: why is this commented?

File: pandas/tests/io/json/test_normalize.py
Comment (jreback) [inline pandas/tests/io/json/test_normalize.py]: need the issue number as a comment

File: pandas/tests/io/json/test_normalize.py
Comment (jreback) [reply pandas/tests/io/json/test_normalize.py]: maybe better to just inline it in the test function

**Review by jreback — CHANGES_REQUESTED** [review]
haven't really looked at the implementation, but has some issues to fix first

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: Because `info.governer` is nested.

File: pandas/io/json/normalize.py
Comment (bhavaniravi) [reply pandas/io/json/normalize.py]: This is a mistake. Default is `None`. @WillAyd mentioned in one of the old comments

**Comment by bhavaniravi** [comment]
Not sure why this test case `test_missing` is failing. @WillAyd Can you help me figure this out. https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=10829

**Comment by WillAyd** [comment]
Try merging master one more time - this issue was just fixed recently.

File: pandas/io/json/normalize.py
Comment (WillAyd) [inline pandas/io/json/normalize.py]: Can you revert the change to this line?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Sorry didn't see latest updates on the review side. I still really think it would be better if we just went forward with `max_level` here and did `ignore_keys` as a follow up - otherwise we've kind of implicitly bundled this together which slows down reviews and makes code more error prone.

Do you have any objection to splitting these up into separate PRs? If not can you focus on `max_level` in this one and add a whatsnew note?

**Comment by bhavaniravi** [comment]
> Sorry didn't see latest updates on the review side. I still really think it would be better if we just went forward with `max_level` here and did `ignore_keys` as a follow up - otherwise we've kind of implicitly bundled this together which slows down reviews and makes code more error prone.
> 
> Do you have any objection to splitting these up into separate PRs? If not can you focus on `max_level` in this one and add a whatsnew note?

@WillAyd I understand. Should I close this and create 2 new PRs. I'm just not sure how to proceed from here.

**Comment by WillAyd** [comment]
You can keep this PR and just focus on the max_level piece - can come back to the second enhancement after that in another PR

**Comment by bhavaniravi** [comment]
@WillAyd sure, let's do that. I have a whole weekend ahead of me. So what's next?

**Comment by WillAyd** [comment]
@bhavaniravi try stripping out anything that isn't relevant for `max_level` to work

**Comment by WillAyd** [comment]
@bhavaniravi any interest in continuing here? I think these changes are great just need to simplify a bit to get through. Let me know if you can address comments above

**Comment by bhavaniravi** [comment]
@WillAyd Moved the `max_level` to a different PR. Will give one for `ignore_keys` after it gets merged.

File: pandas/io/json/normalize.py
Comment (jreback) [inline pandas/io/json/normalize.py]: can u type these parameters

**Review by jreback — CHANGES_REQUESTED** [review]
will have to take a look

**Comment by bhavaniravi** [comment]
Closing this we will track it in 2 separate PRs

**Pull request closed without merging**
